File: app/src/main/java/app/islammedia/halaqatalquran/Database/Converters.java
package app.islammedia.halaqatalquran.Database;

import androidx.room.TypeConverter;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;

public class Converters {

    @TypeConverter
    public static ArrayList<Integer> ArrayListOfIntegersFromString(String value) {
        Type listType = new TypeToken<ArrayList<Integer>>() {}.getType();
        return new Gson().fromJson(value, listType);
    }

    @TypeConverter
    public static String StringFromArrayListOfInteger(ArrayList<Integer> list) {
        Gson gson = new Gson();
        return gson.toJson(list);
    }

    @TypeConverter
    public static ArrayList<String> ArrayListOfStringsFromString(String value) {
        Type listType = new TypeToken<ArrayList<String>>() {}.getType();
        return new Gson().fromJson(value, listType);
    }

    @TypeConverter
    public static String StringFromArrayListOfStrings(ArrayList<String> list) {
        Gson gson = new Gson();
        return gson.toJson(list);
    }

    @TypeConverter
    public static Calendar fromLong(Long l) {
        if (l==null)
            return null;

        Calendar c = Calendar.getInstance();
        c.setTime(new Date(l));
        return c;
    }

    @TypeConverter
    public static Long fromCalendar(Calendar c) {
        if (c==null)
            return null;

        return c.getTimeInMillis();
    }
}

File: app/src/main/java/app/islammedia/halaqatalquran/database/Converters.kt
package app.islammedia.halaqatalquran.database

import android.util.Log
import androidx.room.TypeConverter
import app.islammedia.halaqatalquran.database.helpers.HalaqaDate
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import java.util.*

object Converters {
    @TypeConverter
    fun ArrayListOfIntegersFromString(value: String?): ArrayList<Int?>? {
        val listType = object : TypeToken<ArrayList<Int?>?>() {}.type
        return Gson().fromJson(value, listType)
    }

    @TypeConverter
    fun StringFromArrayListOfInteger(list: ArrayList<Int?>?): String? {
        val gson = Gson()
        return gson.toJson(list)
    }

    @TypeConverter
    fun ArrayListOfStringsFromString(value: String?): ArrayList<String?>? {
        val listType = object : TypeToken<ArrayList<String?>?>() {}.type
        return Gson().fromJson(value, listType)
    }

    @TypeConverter
    fun StringFromArrayListOfStrings(list: ArrayList<String?>?): String? {
        val gson = Gson()
        return gson.toJson(list)
    }

    @kotlin.jvm.JvmStatic
    @TypeConverter
    fun fromLong(l: Long?): Calendar? {
        if (l == null) return null
        val c = Calendar.getInstance()
        c.time = Date(l)
        return c
    }

    @kotlin.jvm.JvmStatic
    @TypeConverter
    fun fromCalendar(c: Calendar?): Long? {
        return c?.timeInMillis
    }

    @kotlin.jvm.JvmStatic
    @TypeConverter
    fun fromHalaqaDates(hds: MutableList<HalaqaDate>): String? {
        if (hds.size == 0) return null
        val hdsString = StringBuilder()
        for (hd in hds) hdsString.append(hd.getDate()?.timeInMillis).append("_").append(hd.getRep()).append("&")
        hdsString.deleteCharAt(hdsString.length - 1)
        return hdsString.toString()
    }

    @kotlin.jvm.JvmStatic
    @TypeConverter
    fun toHalaqaDate(hdsString: String?): MutableList<HalaqaDate?>? {
        if (hdsString == null) return null
        val hdStr: Array<String> = hdsString.split("&".toRegex()).toTypedArray()
        val halaqaDates: MutableList<HalaqaDate?> = ArrayList()
        for (s in hdStr) {
            val calReps: Array<String> = s.split("_".toRegex()).toTypedArray()
            val c = Calendar.getInstance()
            c.timeInMillis = calReps[0].toLong()
            halaqaDates.add(HalaqaDate(c, calReps[1].toInt()))
        }
        Log.i("dates number", halaqaDates.size.toString())
        return halaqaDates
    }
}